docs/conditions.rst

```diff
@@ -132,6 +132,223 @@ to set those conditions inside the init() method of your model::
+
+.. php:method:: mergeAnd(AbstractScope $scopeA, AbstractScope $scopeB, $_ = null);
+
+Merge number of scopes using AND as junction. Returns the resulting scope.
+
+.. php:method:: mergeOr(AbstractScope $scopeA, AbstractScope $scopeB, $_ = null);
+
+Merge number of scopes using OR as junction. Returns the resulting scope.
+
+.. php:method:: merge(AbstractScope $scopeA, AbstractScope $scopeB, $junction = self::AND);
+
+Merge two scopes using $junction. Returns the resulting scope.
+
+.. php:method:: find($key);
+
+Returns array of conditions whose key property matches the $key parameter. Useful when renaming fields and updating saved scope objects
+
+.. php:method:: addComponent(AbstractScope $scope);
+
+Add a component (sub-scope or sub-condition) to the scope. The scope junction (either AND or OR remains same)
+
+.. php:method:: getAllComponents();
+
+Get array of all scope components.
+
+.. php:method:: getActiveComponents();
+
+Get array of only active scope components.
+
+.. php:method:: peel();
+
+Peels off nested scopes with single contained component. Useful for converting (((field = value))) to field = value.
+
+.. php:method:: clear();
+
+Clears the scope from components.
+
+.. php:method:: any();
+
+Checks if scope components are joined by OR
+
+.. php:method:: all();
+
+Checks if scope components are joined by AND
+
+Conditions on Referenced Models
+-------------------------------
+
+Agile Data allows for adding conditions on related models for retrieval of type 'model has references where'.
+
+Setting conditions on references can be done utilizing the Model::refLink method but there is a shorthand format 
+directly integrated with addCondition method using "/" to chain the reference names::
+
+	$contact->addCondition('company/country', 'US');
+	
+This will limit the $contact model to those whose company is in US.
+'company' is the name of the reference in $contact model and 'country' is a field in the referenced model.
+
+If a condition must be set directly on the existence or number of referenced records the special symbol "#" can be
+utilized to indicate the condition is on the number of records::
+
+	$contact->addCondition('company/tickets/#', '>', 3);
+	
+This will limit the $contact model to those whose company have more than 3 tickets.
+'company' and 'tickets' are the name of the chained references ('company' is a reference in the $contact model and
+'tickets' is a reference in Company model)
+
+For applying conditions on existence of records the '?' (has any) and '!' (doesn't have any) special symbols can be used.
+Although it is similar in functionality to checking ('company/tickets/#', '>', 0) or ('company/tickets/#', '=', 0)
+'?' and '!' special symbols use optimized query and are much faster::
+
+	// Contact whose company has any tickets
+	$contact->addCondition('company/tickets/?');
+
+	// Contact whose company doesn't have any tickets
+	$contact->addCondition('company/tickets/!');
+
+Condition Value Placeholder
```

Comment:
The placeholder is important as you save the scope in database.
This means the same scope is saved (using placeholders) but it will be differently applied depending on system user, etc.
E.g if you have a placeholder `__PERSPECTIVE__` which resolves to a list of companies the user is responsible for.
Then you can save a scope for all users but it will resolve to different actual scope when applied to the model based on user presently active.
All these functionalities are refined implementation based on the real use, business proven designs.

Reply:
OK probably placeholders are good too, but still need to think about arguments.

For example, think about scope to select invoices issued last year.
In data alone it can be implemented with placeholders something like this, right?
```
Condition::registerPlaceholder('__FROM__', [
    'label' => 'Date From',
    'value' => function(Condition $condition) {
    	return new \DateTime('first day of last year);
    },
]);
Condition::registerPlaceholder('__TILL__', [
    'label' => 'Date Till',
    'value' => function(Condition $condition) {
    	return new \DateTime('last day of last year);
    },
]);

$condition = Condition::create([
    ['date', '>=', '__FROM__'],
    ['date', '<=', '__TILL__'],
]);
```
If I get these dates from somewhere else then I still can use registerPlaceholder() method to set them.

But what if I need to ask user (in UI) to fill in these placeholder values? We can ask values for all placeholders maybe, but still then we have no knowledge of value types here and UI will never know that it has to ask for date value instead of string.

It's not that I don't like placeholders, but there is something missing about them. Anyway I guess we can extend that later on.